src/Workspaces/Core/Portable/EmbeddedLanguages/DateTime/LanguageServices/DateTimePatternDetector.cs

```diff
@@ -0,0 +1,202 @@
+{
+    /// <summary>
+    /// Helper class to detect <see cref="System.DateTime"/> and <see cref="DateTimeOffset"/> format
+    /// strings in a document efficiently.
+    /// </summary>
+    internal sealed class DateTimePatternDetector
+    {
+        private const string FormatName = "format";
+
+        private static readonly ConditionalWeakTable<SemanticModel, DateTimePatternDetector?> _modelToDetector =
+            new ConditionalWeakTable<global::Microsoft.CodeAnalysis.SemanticModel, DateTimePatternDetector?>();
+
+        private readonly EmbeddedLanguageInfo _info;
+        private readonly SemanticModel _semanticModel;
+        private readonly INamedTypeSymbol _dateTimeType;
+        private readonly INamedTypeSymbol _dateTimeOffsetType;
+
+        public DateTimePatternDetector(
+            SemanticModel semanticModel,
+            EmbeddedLanguageInfo info,
+            INamedTypeSymbol dateTimeType,
+            INamedTypeSymbol dateTimeOffsetType)
+        {
+            _info = info;
+            _semanticModel = semanticModel;
+            _dateTimeType = dateTimeType;
+            _dateTimeOffsetType = dateTimeOffsetType;
+        }
+
+        public static DateTimePatternDetector? TryGetOrCreate(
+            SemanticModel semanticModel, EmbeddedLanguageInfo info)
+        {
+            // Do a quick non-allocating check first.
+            if (_modelToDetector.TryGetValue(semanticModel, out var detector))
+            {
+                return detector;
+            }
+
+            return _modelToDetector.GetValue(
+                semanticModel, _ => TryCreate(semanticModel, info));
+        }
+
+        private static DateTimePatternDetector? TryCreate(
+            SemanticModel semanticModel, EmbeddedLanguageInfo info)
+        {
+            var dateTimeType = semanticModel.Compilation.GetTypeByMetadataName(typeof(System.DateTime).FullName!);
+            var dateTimeOffsetType = semanticModel.Compilation.GetTypeByMetadataName(typeof(System.DateTimeOffset).FullName!);
+            if (dateTimeType == null || dateTimeOffsetType == null)
+                return null;
+
+            return new DateTimePatternDetector(semanticModel, info, dateTimeType, dateTimeOffsetType);
+        }
+
+        public static bool IsDefinitelyNotDateTimeStringToken(SyntaxToken token, ISyntaxFacts syntaxFacts)
+        {
+            if (!syntaxFacts.IsStringLiteral(token))
+                return true;
+
+            if (!IsMethodArgument(token, syntaxFacts))
+                return true;
+
+            return false;
+        }
+
+        private static bool IsMethodArgument(SyntaxToken token, ISyntaxFacts syntaxFacts)
+            => syntaxFacts.IsLiteralExpression(token.Parent) &&
+               syntaxFacts.IsArgument(token.Parent!.Parent);
+
+        public bool IsDateTimeStringToken(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotDateTimeStringToken(token, _info.SyntaxFacts))
```

Comment:
nit- would prefer this to be something like IsStringLiteralOrMethodArgument.

It's a little confusing down below why it can just be assumed that it's a string literal with a parent argument.

Reply:
renamed to IsPossiblyDateAndTimeToken

Reply:
also doc'ed.